File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
from collections import deque


class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        letters, chk = deque(), set()
        result = 0

        for i in range(len(s)):
            if s[i] not in chk:
                letters.append(s[i])
                chk.add(s[i])
            else:
                while letters[0] != s[i]:
                    chk.remove(letters.popleft())
                letters.popleft()
                letters.append(s[i])
            result = max(result, len(letters))

        return result

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Deque<Character> que = new ArrayDeque<>();
        Set<Character> letters = new HashSet<>();
        int result = 0;
        int idx = 0;

        while (idx < s.length()) {
            if (letters.contains(s.charAt(idx))) {
                while (que.peekFirst() != s.charAt(idx)) {
                    letters.remove(que.pollFirst());
                }
                que.pollFirst();
                que.add(s.charAt(idx));
                idx++;
            } else {
                que.add(s.charAt(idx));
                letters.add(s.charAt(idx));
                idx++;
            }
            result = Math.max(result, que.size());
        }
        return result;
    }
}